### ISSUE TITLE
order_cpp needs improvement

### ISSUE THREAD

**Issue by sircosine** [cdowd/twosamples#8]
Hi,

I was looking into speeding up `order_cpp` but stumbled upon a seemingly incorrect behaviour, showing different results from `order()` and `order_cpp()` 

```r
> library(twosamples)
> x = c(9, 7, 8)
> order(x)
[1] 2 3 1
> order_cpp(x) + 1              # add one to correct zero-based C++ indexing
[1] 3 1 2                       # expected order: 2 3 1
> 
```

My suggestion would be to use the STL implementation instead:
```c
IntegerVector order_cpp_stl(NumericVector x) {
  int n = x.size();
  IntegerVector ord(n);
  for (int i = 0 ; i < n ; i++) {
    ord[i] = i;
  }
  // Sort indices using the values from x
  std::sort(ord.begin(), ord.end(),
       [&x](const int& a, const int& b) {
         return (x[a] < x[b]);
       });
  return ord;
}
```
which should not only be correct but also faster:
```r
> z <- rnorm(10000)
> stopifnot(all.equal(order_cpp_stl(z) + 1, order(z)))       # see above order_cpp_stl()
> print(microbenchmark(order_cpp(z), order_cpp_stl(z)), unit="ms", signif=3)
Unit: milliseconds
             expr     min      lq        mean  median     uq     max neval
     order_cpp(z) 325.000 325.000 325.7930555 326.000 326.00 329.000   100
 order_cpp_stl(z)   0.547   0.551   0.5578429   0.556   0.56   0.733   100
```

I am sorry for bugging you again - I didn't mean to distract you from your dissertation, but it would be great to verify the above behaviour and post a quick fix, if confirmed. After you are done with your thesis, we can revisit this and maybe talk about adding some unit tests with `testthat`,`tinytest` or whatever is appropriate and/or lightweight enough.

Thank you!

**Comment by cdowd** [comment]
Okay. At first glance that looked to be a huge problem -- thank you for bringing this up. Honestly, my first reaction after testing and getting the same result as you was "how do any of the tests have any power if the sorting is broken?". 

You are correct, this is not producing the same output as `order` does -- which is a problem. The simplest way of summarizing the situation is the following. Typically you would run code that looks like:
     dd = d[order(d)]  
And that would give a vector, `dd`, which is sorted. 
For reasons surpassing my understanding now (two years later), I decided I wanted the following code to return a sorted vector:
     dd[order(d)] = d
I believe you'll find that `order_cpp` as it stands, works for that -- and the code I wrote using `order_cpp` uses that method for producing a sorted vector. I honestly can't defend the reasoning behind producing a sorted vector in this manner -- and this should be fixed before a new release, but the current code isn't broken. 

You are right that the sort code is slow. Very slow. And the STL would improve the speed. This is... relatively high priority. I had thought I was using a much faster sorting algorithm, but I was mistaken. This is an O(N^2) sort, slowing the whole codebase down substantially. Unfortunately, so far I can't seem to get the code you suggested working on my machine, nor have I managed to make any other of the relatively straightforward suggestions on StackOverflow work. I will keep you apprised.

**Comment by cdowd** [comment]
At the moment, I have a working sort algorithm based on [this](https://www.geeksforgeeks.org/sorting-vector-of-pairs-in-c-set-1-sort-by-first-and-second/). It is below. 

    IntegerVector order_stl(NumericVector x) {
       int n = x.size();
       std::vector<std::pair<double,int> > paired;
       for(int i=0;i<n;i++){
         std::pair<double,int> P = std::make_pair(x[i],i);
         paired.push_back(P);
       }
       sort(paired.begin(),paired.end());
       IntegerVector indices(n);
       for (int i=0;i<n;i++){
          indices[i] = paired[i].second;
       }
       return indices;
     }


It is much faster than the current iterations -- as it uses the STL for its sort. But the pair creation seems to be slowing it down considerably -- probably because it needs to expand the memory N times. Working on changing that. This is (like your suggestion above) not a drop-in replacement. all entries of `dd[order(d)] = d` will need to switch to `dd = d[order(d)]` for this to work.

**Comment by cdowd** [comment]
// [[Rcpp::export]]
    IntegerVector order_stl_2(NumericVector x) {
      int n = x.size();
      std::vector<std::pair<double,int> > paired;
      paired.reserve(n);
      for(int i=0;i<n;i++){
        paired[i] = std::make_pair(x[i],i);
      }
      sort(paired.begin(),paired.end());
      IntegerVector indices(n);
      for (int i=0;i<n;i++){
        indices[i] = paired[i].second;
      }
      return indices;
    }

Seems to have resolved the timing issue. I'll update the entire package here (on github) shortly. Testing and deployment to CRAN is... costly, so I'll probably wait a few weeks before doing that.

**Comment by sircosine** [comment]
Thanks a lot for clarifying the input vs. permutation indexing, that seemed a lot less than obvious to me :)

Another alternative is maybe to leave all those `dd[order(d)] = d` assignments as they are, and make `order_cpp` produce a permutation index instead (though, personally, I would find `order()`-equivalent approach simpler and more intuitive):

```c
//////////////////////////// C++ 98 ////////////////////////////
IntegerVector order_cpp_c98(NumericVector x) {
  int n = x.size();
  std::vector<std::pair<double, int> > paired(n);
  for (int i = 0; i < n; i++) {
    paired[i] = std::make_pair(x[i], i);
  }
  std::sort(paired.begin(), paired.end());
  IntegerVector indices(n);
  for (int i = 0; i < n; i++) {
    indices[paired[i].second] = i;
  }
  return indices;
}

//////////////////////////// C++ 11 ////////////////////////////
IntegerVector order_cpp_stl(NumericVector x) {
  int n = x.size();
  IntegerVector ord(n);
  for (int i = 0 ; i < n ; i++) {
    ord[i] = i;
  }
  std::sort(ord.begin(), ord.end(),
       [&x](const int& a, const int& b) {
         return (x[a] < x[b]);
       });
  IntegerVector permutation_ord(n);
  for (int i = 0 ; i < n ; i++) {
    permutation_ord[ord[i]] = i;
  }
  return permutation_ord;
}
```
 
I tried both C++ 11 and C++ 98 (with `std::pair`s), they seems to have comparable performance:
```r
> z <- rnorm(10000)
> stopifnot(all.equal(order_cpp_stl(z), order_cpp(z)))
> stopifnot(all.equal(order_cpp_c98(z), order_cpp(z)))
> print(microbenchmark(order_cpp(z), order_cpp_stl(z), order_cpp_c98(z)), unit="ms", signif=3)
Unit: milliseconds
             expr     min      lq        mean  median      uq     max neval
     order_cpp(z) 324.000 324.000 324.6438917 325.000 325.000 328.000   100
 order_cpp_stl(z)   0.561   0.564   0.6284115   0.568   0.574   6.170   100
 order_cpp_c98(z)   0.576   0.587   0.6012051   0.594   0.599   0.682   100
```

Thanks!

**Comment by cdowd** [comment]
Okay. Your c98 rendition was the fastest I've seen, so it is what I'm using for now. I seem to have some kind of ongoing issue with lambdas which I'll maybe sort out another day. I've fixed the functions here on github. I'll wait a few weeks before trying to post the fix to CRAN, as it is costly and I want some time to notice new bugs first. 

As for retaining the old order method... I much prefer to maintain consistency with R's `base::order()`, so I made the slight edit necessary for that -- hopefully without causing issues. When I post to CRAN, I'll change the names of everything in order to deprecate the old function 'safely', and at that point I'll close this issue. 

I really appreciate your spotting this. Thanks again.
